go/vt/vtgate/healthcheck.go

```diff
@@ -0,0 +1,279 @@
+		if err != nil {
+			return err
+		}
+		stream, errfunc, err = conn.StreamHealth(ctx)
+		if err != nil {
+			conn.Close()
+			return err
+		}
+		err = processResponse()
+		if err != nil {
+			conn.Close()
+			return err
+		}
+		hcc.mu.Lock()
+		hcc.conn = conn
+		hcc.mu.Unlock()
+		return nil
+	}
+
+	// main loop to retry health check if it fails
+	for {
+		err = connectFunc()
+		if err != nil {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			log.Errorf("cannot connect to %+v: %v", endPoint, err)
+			time.Sleep(hc.retryDelay)
+			continue
+		}
+		for {
+			err = processResponse()
+			if err != nil {
+				conn.Close()
+				select {
+				case <-ctx.Done():
+					return
+				default:
+				}
+				log.Errorf("error when streaming tablet health from %+v: %v", endPoint, err)
+				time.Sleep(hc.retryDelay)
+				break
+			}
+		}
+	}
+
+}
+
+// onAddrUpdate handles endpoint up/down events.
+func (hc *HealthCheck) onAddrUpdate(updates <-chan *AddrUpdate) {
+	for update := range updates {
+		if update.IsUp {
+			go hc.checkConn(update.EndPoint, nil)
+			continue
+		}
+		addr := hc.endPointToString(update.EndPoint)
+		hc.mu.RLock()
+		hcc := hc.addrToConns[addr]
+		hc.mu.RUnlock()
+		hcc.cancelFunc()
+	}
+}
+
+// endPointToString builds addr string from host and ports.
+func (hc *HealthCheck) endPointToString(endPoint *pbt.EndPoint) string {
+	parts := make([]string, 0, 1)
+	for name, port := range endPoint.PortMap {
+		parts = append(parts, name+":"+string(port))
+	}
+	sort.Strings(parts)
+	parts = append([]string{endPoint.Host}, parts...)
+	return strings.Join(parts, ":")
+}
+
+// addEndPointToTarget adds the endpoint to the given target.
+func (hc *HealthCheck) addEndPointToTarget(target *pbq.Target, endPoint *pbt.EndPoint) {
+	shardMap, ok := hc.targetToEPs[target.Keyspace]
+	if !ok {
+		shardMap = make(map[string]map[pbt.TabletType][]*pbt.EndPoint)
+		hc.targetToEPs[target.Keyspace] = shardMap
+	}
+	ttMap, ok := shardMap[target.Shard]
+	if !ok {
+		ttMap = make(map[pbt.TabletType][]*pbt.EndPoint)
+		shardMap[target.Shard] = ttMap
+	}
+	epList, ok := ttMap[target.TabletType]
+	if !ok {
+		epList = make([]*pbt.EndPoint, 0, 1)
+		ttMap[target.TabletType] = epList
+	}
+	for _, ep := range epList {
+		if topo.EndPointEquality(ep, endPoint) {
+			log.Warningf("endpoint is already up: %+v", endPoint)
+			return
+		}
+	}
+	ttMap[target.TabletType] = append(epList, endPoint)
+}
+
+// getEndPointFromTarget returns all endpoints for the given keyspace/shard.
+func (hc *HealthCheck) getEndPointFromTarget(keyspace, shard string) []*pbt.EndPoint {
+	shardMap, ok := hc.targetToEPs[keyspace]
+	if !ok {
+		return nil
+	}
+	ttMap, ok := shardMap[shard]
+	if !ok {
+		return nil
+	}
+	res := make([]*pbt.EndPoint, 0, 1)
+	for _, epList := range ttMap {
+		res = append(res, epList...)
+	}
+	return res
+}
+
+// deleteEndPointFromTarget deletes the endpoint for the given target.
```

Comment:
For all methods that assume the lock is already taken, please ad a comment to that effect (and maybe even add a 'Protected' suffix or something like that)